### PR TITLE
Change static fields emits

### LINKED ISSUE #36267: Referencing `this` from a static class field
**TypeScript Version:**  Nightly Playground

The following code ([Playground link](http://www.typescriptlang.org/play/?ts=Nightly#code/MYGwhgzhAEDC0G8BQ1oQC5nQS2NM0AvNAORgkDcKamOeARkdOgBbYQB0YVAvksAHsAdhAEgAphxACA5gApYHegEoK0APTrS5IA)) is legal in JavaScript, but produces a static error in TypeScript:

```ts
class C {
  static a = 'a';
  static b = this.a;
    // 'this' cannot be referenced in a static property initializer.
}
console.log(C.b); // 'a'
```

Practical use case – [my library Enumify for enums](https://github.com/rauschma/enumify):

```js
class Color extends Enumify {
  static red = new Color();
  static orange = new Color();
  static yellow = new Color();
  static green = new Color();
  static blue = new Color();
  static purple = new Color();
  static _ = this.closeEnum(); // TypeScript: Color.closeEnum()
}
```

### PR DESCRIPTION
**Pull request by Kingwl**
Fixes #36267 

TODO:

- [x] Update transformer.
  - [x] transform this
  - [x] transform super
  - [x] esnext, es6, es5   
  - [x] useDefineForClassFields 
- [x] Avoid errors in checker.
  - [x] `this` in static context
  - [x] `super` in static context
- [x] Add more tersts.
  - [x] this
  - [x] super
  - [x] new this
  - [x] super call
  - [x] complex expression
  - [x] function boundary
  - [x] arrow function boundary
  - [x] class boundary

### CODE REVIEW

**Comment by Kingwl** [comment]
@rbuckton Sorry for bothering. But take a look please.

**Comment by sandersn** [comment]
Before I look at the PR, I remember talking to @rbuckton about this last year and getting confused about how to make decorators work after he pointed out that they could be troublesome. Can you add some tests with decorated classes?
(or point me to the existing ones, although I didn't see any filenames with 'decorator' in them).

**Comment by Kingwl** [comment]
Added a basic test.

I think an acceptable solution is disallow static this reference in a class who has decorator. (It's just stage 2 :XD)

**Comment by Kingwl** [comment]
I think we could just replace this to some variable, the same as  the class name reference:

```ts
@foo
class C {
  static a = 1
  static b = () => C.b
}
```

to 

```ts
var C_1;
let C = C_1 = class C {
};
C.a = 1;
C.b = () => C_1.b;
C = C_1 = __decorate([
    foo
], C);
```

**Comment by rbuckton** [comment]
`this` and Decorators *is* confusing. The direction that the decorators proposal seems to be going is that static initializers (field init and `static {}`) will be applied *before* decorators are run since a class decorator acts on a fully evaluated `class`. So `this` would refer to the *undecorated* class (and so would `C`, which differs from TypeScript today). If you wanted to refer to the *decorated* class (if a decorator performed constructor replacement), you would need to do something like this:

```js
let C = @decorator class { 
  static x = 1;
  static y = this.x; // refers to undecorated class
  static z = C.y; // TDZ error since C hasn't been initialized yet.
  static w = () => C.y; // Ok if called after C is initialized.
}
```

**Comment by Kingwl** [comment]
So I guess the correct emit should be:

```ts

@foo
class C {
  static a = 1
  static b = 1 + C.a
  static c = () => C.b
  static d = () => this.b
}

```


```ts
var C_1, C_2;
let C = C_1 = C_2 = class C {
};
C.a = 1;
C.b = 1 + C_1.a;
C.c = () => C_1.b;
C.d = () => C_2.b;
C = C_1 = __decorate([
    foo
], C);
```

BTW: Seems we does not support decorator on class expression...

**Comment by rbuckton** [comment]
I discussed this with the TC39 Decorator stakeholders group. There's no final answer on how class decorators should interact with `this` in static initializers at this time. For now it would be best to just ban the combination and report an error if you use a class decorator on a class with a static initializer that uses `this`. I'd be tempted to say we should emit `(void 0)` in place of `this` when encountering this combination during transformation so people don't just `// @ts-ignore` the error and rely on broken behavior.

**Comment by Kingwl** [comment]
Thanks!

**Comment by rbuckton** [comment]
By the way, static private fields and static methods and accessors just landed in master, so you will want to rebase.

**Comment by Kingwl** [comment]
Oh. I just realized. Should we error and avoid the `super` keyword in static class field initializer?

**Comment by rbuckton** [comment]
`super()` would be disallowed, but not `super.x`. This works today in v8:

```js
class A { static x = 1 }
class B extends A { static y = super.x }
B.y; // 1
```

**Comment by Kingwl** [comment]
Sorry for confused. I means with decorated class.

**Comment by Kingwl** [comment]
@typescript-bot pack this.

**Comment by rbuckton** [comment]
> Sorry for confused. I means `super.x` with decorated class.

Yes, `super.x` should be an error in a decorated class for now.

**Comment by Kingwl** [comment]
There's another [problem](https://www.staging-typescript.org/play?ts=4.3.0-dev.20210325#code/MYGwhgzhAEDC0G8BQ1XQgFzBglsaATgKZgAmA9gHYgCe0wAXNAEQBm55z0AvC+5yjSZseaACMwBHqlCQYASUqUiU5GnXosufAG0MACxwQAdMAC60gIwAmAMwBuQRuh7DJ81buP1AXyR+gA):

```ts
class C {
    static readonly c: "foo" = "foo"
    static bar =  class Inner {
        static [this.c] = 123;
        [this.c] = 123; // 
    }
}

```

image


But I don't think it should be include in this PR.

**Comment by rbuckton** [comment]
> But I don't think it should be include in this PR.

That's fine.

File: src/compiler/transformers/classFields.ts
Comment (rbuckton) [inline src/compiler/transformers/classFields.ts:199]: I would add a comment that this will be visited by the caller to `visitorThisOrSuperInStaticFieldContext`. I would note that I'd generally prefer a single walk of the tree, as opposed to visiting a subtree one way and then visiting it again a second time.

File: src/compiler/utilitiesPublic.ts
Comment (rbuckton) [inline src/compiler/utilitiesPublic.ts:1245]: This is used by a transformer. Transformer's shouldn't rely on `parent` pointers. Please see my note in *classFields.ts* about an alternative approach to this.

**Review by rbuckton — CHANGES_REQUESTED** [review]
I still need to review the tests, but wanted to provide additional feedback as soon as possible.

**Comment by rbuckton** [comment]
I've looked at the tests and they seem fine.

File: src/compiler/types.ts
Comment (Kingwl) [reply src/compiler/types.ts:6508]: Would you mind tell me Is the super.xxx(property access) has the flag too? sorry I'm on mobile phone.

Comment (Kingwl) [reply src/compiler/types.ts:6508]: Oh I see the below comments. Never mind.

**Comment by rbuckton** [comment]
If it makes it easier we can drop the `(void 0)` replacement for `this/super` + decorators emit. Its a nice-to-have, but not a requirement since we report an error.

**Comment by Kingwl** [comment]
Would you like to include this one into the beta build today？ If so, i can do these changes in 2 hours.

**Comment by Kingwl** [comment]
Oh we may need further test around the static members. especially about method and accessor？

**Comment by rbuckton** [comment]
> Would you like to include this one into the beta build today？ If so, i can do these changes in 2 hours.

Unfortunately, per @DanielRosenwasser beta cutoff was at 4pm PDT, so this will likely wait until 4.4. I'll let @DanielRosenwasser weigh in, though.

**Comment by DanielRosenwasser** [comment]
I'll be doing the actual branch changes on Saturday, so if the emit is reasonable, we can do that. I'll try to take a look tomorrow.

**Comment by Kingwl** [comment]
I'll do these changes soon(after I get up).

**Comment by Kingwl** [comment]
Sorry. I think we need more time on this. @DanielRosenwasser Please feel free to cutoff.

**Comment by Kingwl** [comment]
Sorry. I have two concerns:

1. `visitorThisOrSuperInStaticFieldContext` has come parameters: `receiver` and `baseClass`. If we visit `this` and `super` with a single walk visitor. Should we store them as a global variables too?

2. `isThisOrSuperInStaticFieldBoundary` need to know the parent. (to check current node is inside the initializer of property rather than the computed name). Should we check them when the visitor reachedevery possible node? Or I'm misunderstanding about that.

File: src/compiler/transformers/classFields.ts
Comment (Kingwl) [reply src/compiler/transformers/classFields.ts]: I think we cannot. otherwise the unnecessary emitted `capture this declaration` could be acceptable.

**Comment by Kingwl** [comment]
PTAL.

**Comment by Kingwl** [comment]
@typescript-bot pack this.

**Comment by Kingwl** [comment]
Upppp

**Comment by Kingwl** [comment]
Up again(

File: src/compiler/transformers/classFields.ts
Comment (rbuckton) [inline src/compiler/transformers/classFields.ts:226]: Is this something we could do in substitution, instead of walking the tree to every `this` or `super` reference during transformation?

You could use `onEmitNode` and enable emit notification to keep track of the outer tree state as the printer walks the tree.

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: src/compiler/transformers/classFields.ts
Comment (Kingwl) [reply src/compiler/transformers/classFields.ts:226]: Sorry, I'm not familiar with `PipelinePhase` (as well as `Substitution`). Are there any docs or introductions about that?

Comment (Kingwl) [reply src/compiler/transformers/classFields.ts:226]: And we a context called staticFieldInitializerContext. And we have visited property declaration.

Do we still need them if we are using `Substitution` ?

Comment (rbuckton) [reply src/compiler/transformers/classFields.ts:226]: A tree walk is an expensive operation, especially when we must perform multiple successive tree walks for each transformation (i.e., `ts` -> `esnext` -> `es2020` -> `es2019` -> `es2018` -> etc.). In order to reduce this cost, we build up `TransformFlags` as we construct an AST that provide information about a node and its subtree.

However, some transformations generally would require descending all the way into every `Identifier` or `ThisExpression` in an AST, which can be costly. For example, converting `this` to `_this` in transformed arrow functions, or `foo` into `some_module_1.foo` for an import, etc. To avoid that expensive walk, we have a separate transformation phase called "Just-in-Time" (JIT) substitution. JIT substitution is invoked by the emitter when the entire AST is walked as we print, since we're guaranteed to visit every node as we descend through the tree.

JIT substitution is broken down into two components:
- Emit Notification
- Substitution

## Emit Notification

With any regular transformation, we generally use the call stack to save and restore context information relevant to children, such as with `staticFieldInitializerContext`:

```ts
function visitSomeNode(node: SomeNode) {
    // save old state as we descend
    const savedState = someState;
    // set new state for subtree
    someState = ...;
        
    const newNode = factory.updateSomeNode(node, ...);
        
    // restore old state
    someState = savedState;

    return newNode;
}
```

When we perform JIT substitution, we often need the same capability. To accommodate this need, the emitter will trigger "emit notifications" that can be set up via the `TransformationContext`:

```ts
export function transformSomething(context: TransformationContext) {
    const previousOnEmitNode = context.onEmitNode; // capture existing notification hook
    context.onEmitNode = onEmitNode; // overwrite the notification hook

    ...

    function onEmitNode(hint: EmitHint, node: Node, emitCallback: (hint: EmitHint, node: Node) => void) {
        // save old state as we descend
        const savedState = someState;
        // set new state for subtree
        someState = ...;

        // re-enter the emitter
        emitCallback(hint, node);

        // restore old state
        someState = savedState;
    }
}
```

Emit notification doesn't necessarily trigger for all `Node`s, however, since that too would be expensive. Instead, you must inform the emitter than emit notification is necessary either for any node with a specific `SyntaxKind`, or by setting an `EmitFlag` on a specific node:

```ts
export function transformSomething(context: TransformationContext) {
    const previousOnEmitNode = context.onEmitNode; // capture existing notification hook
    context.onEmitNode = onEmitNode; // overwrite the notification hook
    context.enableEmitNotification(SyntaxKind.SomeNode);
    ...
}

// or
setEmitFlags(node, EmitFlags.AdviseOnEmitNode);
```

If you need to share state between a regular transformation and a later substitution, we generally do this by creating something like a `Map` in the transformer, and using that to store information for later reuse by using the original node id:

```ts
export function transformSomething(context: TransformationContext) {
    const previousOnEmitNode = context.onEmitNode; // capture existing notification hook
    context.onEmitNode = onEmitNode; // overwrite the notification hook
    context.enableEmitNotification(SyntaxKind.SomeNode);

    let someState: SomeState | undefined;
    const someStateForNode = new Map<number, SomeState>();

    ...

    function visitSomeNode(node: SomeNode) {
        const savedState = someState;
        someState = ...;

        const newNode = factory.updateSomeNode(node, ...);

        // persist state for later substitution
        someStateForNode.set(getOriginalNodeId(node), someState);

        someState = savedState;
        return newNode;
    }

    ...

    function onEmitNode(hint: EmitHint, node: Node, emitCallback: (hint: EmitHint, node: Node) => void) {
        if (node.kind === SyntaxKind.SomeNode) {
            const savedState = someState;
            someState = someStateForNode.get(getOriginalNodeId(node));

            // re-enter the emitter
            emitCallback(hint, node);

            // restore old state
            someState = savedState;
        }
        else {
            emitCallback(hint, node);
        }
    }
}
```

This essentially allows us to restore the necessary surrounding context for a substitution, though its important to keep in mind that the `Node` we encounter during `onEmitNode` may not be the same `Node` we saw during transformation due to subsequent transformations (i.e., an `ArrowFunction` may have been transformed into a `FunctionExpression`). This is why we use `getOriginalNodeId`, since this gets the `id` of the "original" `Node`, a pointer that is set when performing an `factory.updateX` function or manually using `setOriginalNode`.

## Substitution

Once you have established the correct context via emit notification, you can set inject a substitution hook in the same way as an emit notification hook:

```ts
export function transformSomething(context: TransformationContext) {
    const previousOnSubstituteNode = context.onSubstituteNode; // capture existing substitution hook
    context.onSubstituteNode = onSubstituteNode; // overwrite the substitution hook
    context.enableSubstitution(SyntaxKind.ThisKeyword);

    ...

    function onSubstituteNode(hint: EmitHint, node: Node) {
        // first, evaluate existing substitutions
        node = previousOnSubstituteNode(hint, node);

        // now perform our own
        if (node.kind === SyntaxKind.ThisKeyword) {
            return substituteThisExpression(node as ThisExpression);
        }

        return node;
    }

    function substituteThisExpression(node: ThisExpression) {
        // use the state we established via emit notification
        ...
    }
}
```

Comment (Kingwl) [reply src/compiler/transformers/classFields.ts:226]: Thanks for the explanation.
In my rough understanding, this is looks like merging multiple AST traversals into once?

Comment (Kingwl) [reply src/compiler/transformers/classFields.ts]: Another problem is what if the node has already been transformed.
For example:

```ts
class C {
    static a = 1
}

```

to

```ts
var C = /** @class */ (function () {
    function C() {
    }
    C.a = 1;
    return C;
}());

```

Should I enable `Emit Notification` and `Substitute` for the `FunctionDeclaration` or `ClassDeclaration`?

File: src/compiler/transformers/classFields.ts
Comment (rbuckton) [inline src/compiler/transformers/classFields.ts:1214]: Is this one necessary? Shouldn't `currentStaticFieldInitializerContext` already be set as you descend? I'm not sure why you'd need to reset it.

File: tests/baselines/reference/thisInArrowFunctionInStaticInitializer1.js
Comment (rbuckton) [inline tests/baselines/reference/thisInArrowFunctionInStaticInitializer1.js:18]: What happens if `foo` contains a variable named `Vector`? You might want to just generate a temp variable to hold the reference.

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: src/compiler/transformers/classFields.ts
Comment (Kingwl) [reply src/compiler/transformers/classFields.ts]: Friendly ping. needs some suggestions here😂.

Comment (rbuckton) [reply src/compiler/transformers/classFields.ts]: It may not be necessary. You could wire up an `original` pointer to the original PropertyDeclaration on the `C.a = 1` expression using `setOriginalNode`, as well as call `addEmitFlags(node, EmitFlags.AdviseOnEmitNode)`. That will cause emit notification to be triggered for that specific node. Then in your `onEmitNode` hook you can look test for the property declaration:

```ts
function onEmitNode(hint, node, emitCallback) {
  if (isAssignmentExpression(node, true)) {
    const original = getOriginalNode(node);
    if (isPropertyDeclaration(original)) {
      // we have the original property declaration.
      ...
    }
  }
}
```

Note that an "original" node is not guaranteed to be a "parse tree node", and only "parse tree nodes" are guaranteed to have a `parent`. You _could_ use `getParseTreeNode` instead of `getOriginalNode`, but that means that this would break if we later add a transformation that introduces a synthetic property declaration for any reason. To avoid that, you might want to associate the `staticFieldInitializerContext` in `staticFieldInitializerContextMap` with the node id of each `PropertyDeclaration`, rather than just the class itself.

File: tests/baselines/reference/thisInArrowFunctionInStaticInitializer1.js
Comment (Kingwl) [reply tests/baselines/reference/thisInArrowFunctionInStaticInitializer1.js:18]: Ahhh.. Ok. Currently arrow function does not been a function boundary. My fault.

File: src/compiler/transformers/classFields.ts
Comment (Kingwl) [reply src/compiler/transformers/classFields.ts:1214]: > Is this one necessary?

I guess so. I'd like to bind the visiting `currentStaticFieldInitializerContext` with the this or super expression to directly use when substitution.

File: src/compiler/transformers/classFields.ts
Comment (Kingwl) [reply src/compiler/transformers/classFields.ts]: It's a little bit hard if we are enabled useDefineField. Because It's generated a ObjectDefine call.

File: tests/baselines/reference/thisInArrowFunctionInStaticInitializer1.js
Comment (rbuckton) [reply tests/baselines/reference/thisInArrowFunctionInStaticInitializer1.js:18]: That's not what I meant. Consider this scenario:

```
class Vector {
  static x = () => {
    const Vector = 1;
    return this;
  };
}
```
Since `Vector` is redeclared inside of the static block, you can't use `Vector` to access the constructor as if it were `this`. Instead I would emit the following:
```js
var Vector = (function () {
  function Vector() {
  }
  var _a = Vector;
  Vector.x = function () {
    var Vector = 1;
    return _a; // points to the outer `Vector`
  };
  return Vector;
})();
```

**Comment by rbuckton** [comment]
From a conversation with @RyanCavanaugh, it seems like we want to try to wrap 4.4 features today. I will do what I can to get this PR merged by EOD.

**Comment by rbuckton** [comment]
I'm realizing there's more work we need to do on this PR, specifically in how it currently handles `super`, and how to better integrate it with private fields emit. I will focus on getting `static {}` merged tonight and will take another look at this afterwards.

**Comment by Kingwl** [comment]
> I'm realizing there's more work we need to do on this PR. Yep...

**Comment by rbuckton** [comment]
I have it close to finished, I should be pushing an update to your PR shortly.

What's changed:
- I removed substitution for `super`, since that is more complex and needs to happen at a different level (property access, element access, call, tagged template, assignment, destructuring, pre- and postfix increment and decrement, etc.)
- I changed `super` substitution in ES2015 through ES2021 to use `Reflect.get` and `Reflect.set`. ES5 emit doesn't use this because we capture `_super` in ES5.
- I merged the static fields context and private fields environment into a `ClassLexicalEnvironment`, and cleaned up the `onEmitNode` call.

**Comment by rbuckton** [comment]
I also fixed a minor bug in the private fields emit, related to pre-/post-increment/decrement. As written, `this.#x++` (and `super.x++` in a static initializer) would throw a runtime error if the value being incremented is a `bigint`. I'll probably apply this fix to the SystemJS module emit for pre-/post-increment/decrement once this PR has merged.

**Comment by Kingwl** [comment]
Oh wow. Thanks for your hard working.

**Comment by rbuckton** [comment]
@sandersn, @weswigham can either of you review my changes to this PR? In general I think this is fairly close to ready. I'd like to merge tomorrow morning for the 4.4 beta if possible.

**Comment by Kingwl** [comment]
Well... A huge update.

**Review by weswigham — APPROVED** [review]
Should we have a reserved name error if you attempt to transpile a class/variable named `Reflect`, since it's referenced in emit now?

**Comment by rbuckton** [comment]
> Should we have a reserved name error if you attempt to transpile a class/variable named `Reflect`, since it's referenced in emit now?

Probably, yes. Once I get `this` in `static {}` working I'll look into it in a follow-on PR.

**Comment by rbuckton** [comment]
I've just about finished adding `this` support to `static {}`, should have this updated shortly.

**Comment by rbuckton** [comment]
Something that will need a bit more time to consider after this has merged is how control flow should affect the types of static properties:
```ts
class C {
  static x; // infers type through control-flow
  static {
    this.x = 1;
  }
}
```
This is similar to what we do for `constructor() {}`, except you can have multiple `static {}` blocks. I have a temporary solution in place currently, but I think we will need to rethink control flow in a `class` long term. For example:
```ts
class C {
  static x;
  static { this.x = 1; }
  static y = this.x; // y should have type `number`
  static { this.x++; } // this.x should have type `number`
}
```

**Comment by rbuckton** [comment]
The last update fixes some use-before-def errors. I also added support for strict property initialization for statics, but I think I will actually revert that since it would be a possible breaking change and we haven't had the opportunity to discuss it in design meeting.

**Comment by rbuckton** [comment]
@typescript-bot perf test
@typescript-bot run dt
@typescript-bot test this
@typescript-bot user test this

**Review by rbuckton — APPROVED** [review]
I want to review the various test suite results before merging, but I think this looks good.

**Comment by rbuckton** [comment]
@RyanCavanaugh, depending on the test results this should be ready to merge shortly.

**Comment by rbuckton** [comment]
- User/Docker tests are failing, but it looks like an issue with the test suite. As a result I can't tell if there's any differences due to this PR.
- Only difference in RWC seems to be a minor change in source map emit, but nothing concerning.
- I don't imagine there will be issues with DT since this shouldn't affect declaration files.

Currently I'm just waiting on the perf results.

**Comment by rbuckton** [comment]
Perf tests seem like a negligible change. There's an increase in emit time, but I think that can be addressed next week.

**Comment by rbuckton** [comment]
I will also follow up next week with the reserved name error.

**Comment by Kingwl** [comment]
Thanks!

**Pull request merged**
